homeassistant/components/dsmr_reader/config_flow.py

```diff
@@ -0,0 +1,44 @@
+"""Config flow to configure DSMR Reader."""
+from __future__ import annotations
+
+from typing import Any
+
+from homeassistant import config_entries
+from homeassistant.components.mqtt import MqttServiceInfo
+from homeassistant.data_entry_flow import FlowResult
+
+from .const import DOMAIN
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class DsmrReaderFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle DSMR Reader config flow."""
+
+    VERSION = 1
+
+    async def async_step_import(self, import_config: dict[str, Any]) -> FlowResult:
+        """Import from configuration.yaml and create config entry."""
+        # There's no configuration supported for this integration, so we need to hardcode the user input here
+        return await self.async_step_user(user_input={})
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle user step."""
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+        if not self.hass.services.has_service(domain="mqtt", service="publish"):
+            return self.async_show_form(step_id="user", errors={"base": "mqtt_missing"})
+        if user_input is not None:
+            return self.async_create_entry(title="DSMR Reader", data={})
+
+        return self.async_show_form(step_id="user")
+
+    async def async_step_mqtt(self, discovery_info: MqttServiceInfo) -> FlowResult:
+        """Handle a flow initialized by mqtt discovery."""
+        if self._async_in_progress() or self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+        await self.async_set_unique_id(DOMAIN)
+
+        # Offer to register all sensors whenever we encounter at least one dsmr/# topic
+        return await self.async_step_user()
```

Comment:
If you set `title="DSMR Reader"` in the constructor I think you can drop this since the default likely does what you need https://github.com/home-assistant/core/blob/8af0d91676f118c400037c0354de3e5812cd55d4/homeassistant/helpers/config_entry_flow.py#L112

Reply:
Inheriting from `DiscoveryFlowHandler` seems to make things easier, indeed. I somewhat fear for the readability as you won't be able to see where the MQTT step gets triggered, so I'll add that information in a comment. I hadn't encountered that class yet.

The only somewhat ugly thing I'm seeing, is that I am then required to provide a discovery function for devices (of which there are none here). I'll probably add a lambda instruction to create a simple Future that always returns false, otherwise I'd need to add a useless function just for that. It would have been nicer if that function was optional.

Reply:
I've had to add a new private function due to this. Adding it as a lambda or Future didn't seem to be possible. It's a bit of a shame, but it's not the end of the world either. I think I've cleaned up the implementation as much as possible after that change.